File: src/test/java/modelfind/ModelFinderZ3ApiTest.java
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertNull;
import org.junit.Test;
import org.junit.Ignore;

import fortress.msfol.*;
import fortress.modelfind.*;
import fortress.transformers.*;

import java.util.List;
import java.io.*;
import java.util.Map;

public class ModelFinderZ3ApiTest {
    
    Var p = Term.mkVar("p");
    Var q = Term.mkVar("q");
    Var x = Term.mkVar("x");
    Var socrates = Term.mkVar("socrates");
    
    Sort U = Sort.mkSortConst("U");
    
    FuncDecl Human = FuncDecl.mkFuncDecl("Human", U, Sort.Bool());
    FuncDecl Mortal = FuncDecl.mkFuncDecl("Mortal", U, Sort.Bool());
    
    // Propositional tests
    
    @Test
    public void propSat1() throws IOException {
        Theory theory = Theory.empty()
            .withConstant(p.of(Sort.Bool()))
            .withAxiom(Term.mkAnd(p, p));
        
        ModelFinder finder = ModelFinder.createDefault();
        finder.setTheory(theory);
        
        assertEquals(ModelFinderResult.Sat(), finder.checkSat());
    }
    
    @Test
    public void propUnsat1() throws IOException {
        Theory theory = Theory.empty()
            .withConstant(p.of(Sort.Bool()))
            .withAxiom(Term.mkAnd(p, Term.mkNot(p)));
        
        ModelFinder finder = ModelFinder.createDefault();
        finder.setTheory(theory);
        
        assertEquals(ModelFinderResult.Unsat(), finder.checkSat());
    }
    
    @Test
    public void propUnsat2() throws IOException {
        Theory theory = Theory.empty()
            .withConstants(p.of(Sort.Bool()), q.of(Sort.Bool()))
            .withAxiom(Term.mkNot(Term.mkImp(Term.mkAnd(p, q), q)));
        
        ModelFinder finder = ModelFinder.createDefault();
        finder.setTheory(theory);
        
        assertEquals(ModelFinderResult.Unsat(), finder.checkSat());
    }
    
    @Test
    public void propSat2() throws IOException {
        Theory theory = Theory.empty()
            .withConstants(p.of(Sort.Bool()), q.of(Sort.Bool()))
            .withAxiom(Term.mkNot(Term.mkImp(Term.mkOr(p, q), q)));
        
        ModelFinder finder = ModelFinder.createDefault();
        finder.setTheory(theory);
        
        assertEquals(ModelFinderResult.Sat(), finder.checkSat());
    }
    
    // EUF tests
    
    // FOL tests
    
    @Test
    public void socratesScope3() throws IOException {
        Term premise1 = Term.mkForall(x.of(U), Term.mkImp(Term.mkApp("Human", x), Term.mkApp("Mortal", x)));
        Term premise2 = Term.mkApp("Human", socrates);
        Term conjecture = Term.mkApp("Mortal", socrates);
        
        Theory theory = Theory.empty()
            .withSort(U)
            .withConstant(socrates.of(U))
            .withFunctionDeclarations(Human, Mortal)
            .withAxiom(premise1)
            .withAxiom(premise2)
            .withAxiom(Term.mkNot(conjecture));
            
        ModelFinder finder = ModelFinder.createDefault();
        finder.setTheory(theory);
        finder.setAnalysisScope(U, 3);
        assertEquals(ModelFinderResult.Unsat(), finder.checkSat());
    }
}

File: src/test/scala/modelfind/ModelFinderZ3ApiTest.scala
import org.scalatest._

import fortress.msfol._
import fortress.modelfind._
import fortress.transformers._

class ModelFinderZ3ApiTest extends UnitSuite {
    
    val p = Var("p")
    val q = Var("q")
    val x = Var("x")
    val socrates = Var("socrates")
    
    val U = Sort.mkSortConst("U")
    
    val Human = FuncDecl.mkFuncDecl("Human", U, Sort.Bool)
    val Mortal = FuncDecl.mkFuncDecl("Mortal", U, Sort.Bool)
    
    // Propositional tests
    
    test("prop sat 1") {
        val theory = Theory.empty
            .withConstant(p.of(Sort.Bool))
            .withAxiom(And(p, p))
        
        val finder = ModelFinder.createDefault
        finder.setTheory(theory)
        
        finder.checkSat should be (ModelFinderResult.Sat)
    }
    
    test("prop unsat 1") {
        val theory = Theory.empty
            .withConstant(p.of(Sort.Bool))
            .withAxiom(And(p, Not(p)))
        
        val finder = ModelFinder.createDefault
        finder.setTheory(theory)
        
        finder.checkSat should be (ModelFinderResult.Unsat)
    }
    
    test("prop unsat 2") {
        val theory = Theory.empty
            .withConstants(p.of(Sort.Bool), q.of(Sort.Bool))
            .withAxiom(Not(Implication(And(p, q), q)))
        
        val finder = ModelFinder.createDefault
        finder.setTheory(theory)
        
        finder.checkSat should be (ModelFinderResult.Unsat)
    }
    
    test("prop sat 2") {
        val theory = Theory.empty
            .withConstants(p.of(Sort.Bool), q.of(Sort.Bool))
            .withAxiom(Not(Implication(Or(p, q), q)))
        
        val finder = ModelFinder.createDefault
        finder.setTheory(theory)
        
        finder.checkSat should be (ModelFinderResult.Sat)
    }
    
    // EUF tests
    
    // FOL tests
    
    test("socrates scope 3") {
        val premise1 = Forall(x.of(U), Implication(App("Human", x), App("Mortal", x)))
        val premise2 = App("Human", socrates)
        val conjecture = App("Mortal", socrates)
        
        val theory = Theory.empty
            .withSort(U)
            .withConstant(socrates.of(U))
            .withFunctionDeclarations(Human, Mortal)
            .withAxiom(premise1)
            .withAxiom(premise2)
            .withAxiom(Not(conjecture))
            
        val finder = ModelFinder.createDefault
        finder.setTheory(theory)
        finder.setAnalysisScope(U, 3)
        finder.checkSat should be (ModelFinderResult.Unsat)
    }
}
